### PR TITLE
Implement C++ API torch::nn::MultiMarginLoss.

### PR DESCRIPTION
**Pull request by CarMiranda**
Hi @yf225 , here is the C++ frontend API MultiMarginLoss implementation and tests #27198. Could you review it and tell me if it is okay?

I am not entirely sure I used `c10::optional` correctly, but `options.weight()` resulted in a compilation error, so I went with `options.weight().value()` instead of `value_or()` to follow the logic in `torch.nn._WeightedLoss.register_buffer` (where one can pass a `None` value).

Oh, and are the tests supposed to be skipped or did I do something wrong? I ran `pytest test/test_cpp_api_parity.py -k Loss -v` , and the `L1Loss` test passed but the others were skipped...

Thank you for the review in any case!

### CODE REVIEW

File: torch/csrc/api/include/torch/nn/options/loss.h [hunk 1]
```diff
@@ -27,5 +27,27 @@ struct TORCH_API HingeEmbeddingLossOptions {
...
+struct TORCH_API MultiMarginLossOptions {
+  /// Has a default value of :math:`1`. :math:`1` and :math:`2`
+  /// are the only supported values.
+  TORCH_ARG(int64_t, p) = 1;
+  /// Has a default value of :math:`1`.
+  TORCH_ARG(double, margin) = 1.0;
+  /// A manual rescaling weight given to each
+  /// class. If given, it has to be a Tensor of size `C`. Otherwise, it is
+  /// treated as if having all ones.
+  TORCH_ARG(c10::optional<Tensor>, weight) = c10::nullopt;
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/options/loss.h]: My sincere apologies for the change here: after putting more thoughts into how we should represent an optional tensor in C++ API, I came to realize that we should use this instead:
```suggestion
  TORCH_ARG(Tensor, weight) = Tensor();
```
which would be consistent with the representation in other parts of C++ API such as custom autograd functions.
If we do this, then we won't need to do `options.weight().value()` anymore, and can just do `options.weight()`.

File: torch/csrc/api/src/nn/modules/loss.cpp [hunk 2]
```diff
@@ -32,5 +32,30 @@ Tensor HingeEmbeddingLossImpl::forward(
...
+// ============================================================================
+
+MultiMarginLossImpl::MultiMarginLossImpl(
+    const MultiMarginLossOptions& options_)
+    : options(options_) {
+      reset();
+    }
+
+void MultiMarginLossImpl::reset() {
+  register_buffer("weight", options.weight().value());
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/loss.cpp]: It'd be awesome if we also add the following checks (expressed in Python), to match the Python version better:
```python
        if p != 1 and p != 2:
            raise ValueError("only p == 1 and p == 2 supported")
        assert weight is None or weight.dim() == 1
```

File: torch/csrc/api/include/torch/nn/functional/loss.h [hunk 3]
```diff
@@ -17,6 +17,20 @@ inline Tensor hinge_embedding_loss(
       options.reduction());
 }
 
+inline Tensor multi_margin_loss(
+    const Tensor& input,
+    const Tensor& target,
+    const MultiMarginLossOptions& options = {}) {
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/loss.h]: It would be great to add the following checks (expressed in Python) to match the Python version better:
```python
    if p != 1 and p != 2:
        raise ValueError('only p == 1 and p == 2 supported')
    if weight is not None:
        if weight.dim() != 1:
            raise ValueError('weight must be one-dimensional')
```

**Review by yf225 — COMMENTED** [review]
@CarMiranda Thanks so much for the great work! I left some very minor comments, and we should be able to merge it very soon :D

File: torch/csrc/api/src/nn/modules/loss.cpp (hunk 2 above)
Comment (CarMiranda) [reply torch/csrc/api/src/nn/modules/loss.cpp]: Is there a reason to treat these two errors with `raise` and `assert`? In the next comment, both errors are treated using `raise`. In any case, looking at how other errors are handled, I used the `TORCH_CHECK` macro, I hope that was the right way!

Comment (yf225) [reply torch/csrc/api/src/nn/modules/loss.cpp]: Yes using `TORCH_CHECK` is the right way - `raise` and `assert` are both Python syntax and we should use the C++ equivalent. Thanks a lot for your help!

File: test/cpp/api/modules.cpp [hunk 4]
```diff
@@ -979,6 +979,20 @@ TEST_F(ModulesTest, HingeEmbeddingLoss) {
...
+  auto weight = torch::tensor({0.3, 0.3, 0.4}, torch::kFloat);
+  MultiMarginLoss loss(MultiMarginLossOptions().margin(2).weight(weight));
+  auto input = torch::tensor({{0.2, 0.2, 0.6}, {0.1, 0.8, 0.1}, {0.9, 0.09, 0.01}}, torch::requires_grad());
+  auto target = torch::tensor({2, 1, 0}, torch::kLong);
+  auto output = loss->forward(input, target);
+  auto expected = torch::tensor({0.3056}, torch::kFloat);
+  auto s = output.sum();
+  s.backward();
+
+  ASSERT_TRUE(output.allclose(expected, 1e-04));
```
Comment (yf225) [inline test/cpp/api/modules.cpp]: It seems that this assert is failing on CI:
```
Oct 08 13:47:16 [ RUN      ] ModulesTest.MultiMarginLoss
Oct 08 13:47:16 /var/lib/jenkins/workspace/test/cpp/api/modules.cpp:992: Failure
Oct 08 13:47:16 Value of: output.allclose(expected, 1e-04)
Oct 08 13:47:16   Actual: false
Oct 08 13:47:16 Expected: true
Oct 08 13:47:16 [  FAILED  ] ModulesTest.MultiMarginLoss (0 ms)
```
The easiest way to debug is likely to run locally and see what causes the difference in output vs. expected value

**Review by yf225 — COMMENTED** [review]
Thanks a lot for the changes @CarMiranda! It seems that there is test failure for the newly added tests, we might need to debug it locally to see what causes the test failure.

**Comment by yf225** [comment]
@CarMiranda We can use these steps to run the tests:
- Follow https://github.com/pytorch/pytorch#from-source to build your branch of PyTorch from source.
- To test `test/cpp/api/modules.cpp`, run `./build/bin/test_api --gtest_filter=ModulesTest* --gtest_stack_trace_depth=10 --gmock_verbose=info`.
- To test `test/cpp/api/functional.cpp`, run `./build/bin/test_api --gtest_filter=FunctionalTest* --gtest_stack_trace_depth=10 --gmock_verbose=info`.

**Comment by CarMiranda** [comment]
It looks like the errors came from precision, I just replaced 0.3056 by 0.305556 and that did the trick!
Thank you for your reviews @yf225 !

**Review by yf225 — APPROVED** [review]
Thanks a lot for the awesome work @CarMiranda! I will merge it after CI passes :D
